parlai/core/torch_agent.py

```diff
@@ -1551,47 +1520,59 @@ def match_batch(self, batch_reply, valid_inds, output=None):
                 batch_reply[i]['text_candidates'] = cands
         return batch_reply
 
-    def last_reply(self, use_reply='label'):
+    def self_observe(self, self_message: Message) -> None:
         """
-        Retrieve the last reply from the model.
-
-        If available, we use the true label instead of the model's prediction.
+        Observe one's own utterance.
 
-        By default, batch_act stores the batch of replies and this method
-        will extract the reply of the current instance from the batch.
+        This is used so that the agent can incorporate its own response into
+        the dialogue history after a batch_act. Failure to implement this will
+        result in an agent that cannot hear itself speak
 
-        :param use_label:
-            default true, use the label when available instead of the model's
-            generated response.
+        :param self_message:
+            The message corresponding to the output from batch_act.
         """
-        # if the last observation was the end of an episode,
-        # then we shouldn't use it as history
-        if (
-            use_reply == 'none'
-            or not self.observation
-            or self.observation['episode_done']
-        ):
-            return None
+        use_reply = self.opt.get('use_reply', 'label')
+
+        if self.last_observation is None:
+            raise RuntimeError(
+                "You're self_observing without having observed something. Check if "
+                "you're missing a step in your observe/act/self_observe loop."
+            )
+
+        if self.last_observation['episode_done']:
+            # oh this was the last example in the episode. reset the history
+            self.history.reset()
+            # additionally mark the last observation as invalid
+            self.last_observation = None
+            return
 
-        if use_reply == 'label':
-            # first look for the true label, if we aren't on a new episode
+        if use_reply == 'none':
+            # we're not including our own responses anyway.
+            return
+        elif use_reply == 'label':
```

Comment:
wasn't there an option like 'label else model'?

Reply:
That's really what label is.

Reply:
Looks like `label_else_model` is what we used pre TorchAgent:

```
devfair0237 parlai selfobserve $ ag label_else_model
parlai/agents/legacy_agents/memnn/memnn_v0.py
87:            default='label_else_model',
89:            choices=['none', 'model', 'label', 'label_else_model'],

parlai/agents/legacy_agents/seq2seq/seq2seq_v0.py
305:            default='label_else_model',
307:            choices=['none', 'model', 'label', 'label_else_model'],

parlai/agents/legacy_agents/seq2seq/utils_v0.py
324:    useReplies='label_else_model',
355:            useReplies == 'label_else_model' and len(history['labels']) == 0

parlai/agents/starspace/starspace.py
162:            default='label_else_model',
164:            choices=['none', 'model', 'label', 'label_else_model'],

parlai/core/utils.py
70:    useReplies='label_else_model',
107:            useReplies == 'label_else_model' and len(history['labels']) == 0
```